File: KanbanBoardApi/Models/KanbanTasks/KanbanTaskModels.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using KanbanBoardApi.Entities;
using KanbanBoardApi.Models.Users;
using KanbanBoardApi.ModelValidations;

namespace KanbanBoardApi.Models.KanbanTasks;

public record KanbanTaskInputModel
{
    [Required]
    [MinLength(5, ErrorMessage = "Title cannot be less than 5 characters long.")]
    [MaxLength(100, ErrorMessage = "Title cannot exceed 100 characters.")]
    public string? Title { get; init; }

    [Required]
    [MinLength(5, ErrorMessage = "Description cannot be less than 5 characters long.")]
    [MaxLength(2000, ErrorMessage = "Description cannot exceed 2000 characters.")]
    public string? Description { get; init; }

    [Required]
    [EnumDataType(typeof(KanbanTaskStatus))]
    public KanbanTaskStatus? Status { get; init; }

    [IsValidUserId]
    [Display(Name = "Assigned User Id")]
    public int? AssignedUserId { get; init; }
}

public record KanbanTaskModel
{
    public int Id { get; init; }
    public required string Title { get; init; }
    public required string Description { get; init; }
    public KanbanTaskStatus Status { get; init; }
    public double Priority { get; init; }

    public DateTime? AssignedAt { get; init; }
    public DateTime CreatedAt { get; init; }
    public DateTime UpdatedAt { get; init; }

    public int? AssignedUserId { get; init; }
    public int? CreatedByUserId { get; init; }
    public int? UpdatedByUserId { get; init; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public UserModel? AssignedUser { get; init; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public UserModel? CreatedByUser { get; init; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public UserModel? UpdatedByUser { get; init; }
}

public record IdPriority
{
    public int Id { get; init; }
    public double Priority { get; init; }
}

File: kanban-board-app/src/models/kanbanTaskModels.ts
import { type User } from '@/models/userModels';

const kanbanTaskItemType = 'kanban_task';

const KanbanTaskStatus = {
  None: 0,
  ToDo: 1,
  InProgress: 2,
  Done: 3,
} as const;

const kanbanTaskStatuses = [
  {
    status: KanbanTaskStatus.None,
    title: 'Backlog',
    bgColor: 'bg-gray-200',
    borderColor: 'border-gray-200',
    chipBaseStyle: 'bg-gray-50 border-gray-600/50',
    chipContentStyle: 'text-gray-600',
  },
  {
    status: KanbanTaskStatus.ToDo,
    title: 'To Do',
    bgColor: 'bg-amber-200',
    borderColor: 'border-amber-200',
    chipBaseStyle: 'bg-amber-50 border-amber-600/50',
    chipContentStyle: 'text-amber-600',
  },
  {
    status: KanbanTaskStatus.InProgress,
    title: 'In Progress',
    bgColor: 'bg-blue-200',
    borderColor: 'border-blue-200',
    chipBaseStyle: 'bg-blue-50 border-blue-600/50',
    chipContentStyle: 'text-blue-600',
  },
  {
    status: KanbanTaskStatus.Done,
    title: 'Done',
    bgColor: 'bg-green-200',
    borderColor: 'border-green-200',
    chipBaseStyle: 'bg-green-50 border-green-600/50',
    chipContentStyle: 'text-green-600',
  },
];

type KanbanTaskStatusEnum = (typeof KanbanTaskStatus)[keyof typeof KanbanTaskStatus];

interface KanbanTaskModel {
  title: string;
  description: string;
  status: KanbanTaskStatusEnum | null;
  assignedUserId: number | null;
}

interface KanbanTaskCommon {
  id: number;
  title: string;
  description: string;
  status: KanbanTaskStatusEnum;
  priority: number;

  assignedUserId: number | null;
  createdByUserId: number | null;
  updatedByUserId: number | null;

  assignedUser: User | null;
  createdByUser: User | null;
  updatedByUser: User | null;
}

interface KanbanTaskServer extends KanbanTaskCommon {
  assignedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

interface KanbanTask extends KanbanTaskCommon {
  assignedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

export {
  kanbanTaskItemType,
  KanbanTaskStatus,
  kanbanTaskStatuses,
  type KanbanTaskStatusEnum,
  type KanbanTaskModel,
  type KanbanTaskServer,
  type KanbanTask,
};
